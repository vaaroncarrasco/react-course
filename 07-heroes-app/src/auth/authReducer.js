import { types } from '../types/types';

// * authReducer is just a pure function
export const authReducer = (state = {}, action) => {

    // const state = {
    //     name: 'Fernando',
    //     logged: true
    // }

    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }

        default:
            return state;
    }

}