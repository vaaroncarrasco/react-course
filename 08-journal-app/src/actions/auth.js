import Swal from "sweetalert2";

import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ( { user } ) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

                dispatch( finishLoading() );


            }).catch( err => {
                dispatch( finishLoading() );
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }
}


export const startRegisterWithEmailPasswordName = ( email, password, name ) => {

    // ? async fns return (callbacks) => {}
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async ( {user} ) => {

                await user.updateProfile({ displayName: name });

                console.log(user)


                dispatch(
                    login( user.uid, user.displayName )
                )

            }).catch( err => {
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }

}


export const startGoogleLogin = () => {

    return (dispatch) => {

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ( { user } ) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });

    }

}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {

    // ? Returns an async callback

    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );

        dispatch( noteLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})