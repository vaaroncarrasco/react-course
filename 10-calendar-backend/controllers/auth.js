const { response } = require('express');
const bcrypt = require('bcryptjs');
const Usuario = require('../models/Usuario');
const { generarJWT } = require('../helpers/jwt');

const crearUsuario = async( req, res = response ) => {

    const { email, password } = req.body;

    try {

        let usuario = await Usuario.findOne({ email });

        if ( usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'User already exists'
            });
        }

        usuario = new Usuario( req.body );

        // * encrypt password before saving to db
        const salt = bcrypt.genSaltSync();
        usuario.password = bcrypt.hashSync( password, salt );

        await usuario.save();

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(201).json({
            ok: true,
            msg: 'registro',
            uid: usuario.id,
            name: usuario.name,
            token
        });

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }
}

const loginUsuario = async ( req, res = response ) => {

    const { email, password } = req.body;

    try {

        const usuario = await Usuario.findOne({ email });

        if ( !usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'No user found with email'
            });
        }

        // Confirm pass
        const validPassword = bcrypt.compareSync( password, usuario.password );

        if ( !validPassword ) {
            return res.status(400).json({
                ok: false,
                msg: 'Password incorrecto'
            });
        }

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        })



    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }

}


const revalidarToken = async( req, res = response ) => {

    const { uid, name } = req;

    // * generate new JWT
    const token = await generarJWT( uid, name );

    res.json({
        ok: true,
        token,
        uid,
        name
    });
}

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}