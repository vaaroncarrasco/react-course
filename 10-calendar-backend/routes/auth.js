// * Auth / Users - Route | host + /api/auth

// node uses same lib it has in memory without importing it twice
const { Router } = require('express');
const { check } = require('express-validator');
const { validarCampos } = require('../middlewares/validar-campos');
const { validarJWT } = require('../middlewares/validar-jwt');
const router = Router();


const {
    crearUsuario,
    loginUsuario,
    revalidarToken
} = require('../controllers/auth');


router.post(
    '/new',
    [ // middlewares
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    crearUsuario
);

router.post(
    '/',
    [ // middlewares
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    loginUsuario
);

router.get('/renew', validarJWT, revalidarToken);

module.exports = router;