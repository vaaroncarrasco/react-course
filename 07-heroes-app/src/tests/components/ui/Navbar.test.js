import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Router } from 'react-router';
import '@testing-library/jest-dom'
import { AuthContext } from '../../../auth/AuthContext';
import { Navbar } from '../../../components/ui/Navbar';
import { types } from '../../../types/types';

describe('testing <Navbar />', () => {

    // ? Testing history w/ history mock
    const historyMock = { // mocking history{}
        location: {},
        push: jest.fn(),
        listen: jest.fn(),
        replace: jest.fn(),
        createHref: jest.fn(),
    }

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name: 'EL PEPE'
        }
    }

    const wrapper = mount(
        <AuthContext.Provider value={ contextValue }>
            <MemoryRouter>

                {/* // ? Router history prop is accessed by child components */}
                <Router history={ historyMock }>

                    <Navbar />

                </Router>

            </MemoryRouter>
        </AuthContext.Provider>
    );

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('should display correctly', () => {
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.text-info').text().trim() ).toBe('EL PEPE');
    });

    test('should call logout and use history', () => {
        wrapper.find('button').prop('onClick')();

        // ? Test dispatch
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.logout
        });

        // ? Test useHistory() -> history.repalce('/');
        expect( historyMock.replace ).toHaveBeenCalledWith('/login');
    });

});