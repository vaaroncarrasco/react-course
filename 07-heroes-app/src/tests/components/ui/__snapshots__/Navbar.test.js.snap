// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <Navbar /> should display correctly 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Heroes App
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-time nav-link text-info"
      >
        EL PEPE
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
